daemon/algod/api/server/v2/test/handlers_test.go

```diff
@@ -693,21 +696,374 @@ func TestPostTransaction(t *testing.T) {
+		},
+		{
+			txnIndex:                   0,
+			format:                     "msgpack",
+			expectedStatus:             404,
+			enableTransactionSimulator: false,
+		},
+		{
+			txnIndex:                   0,
+			format:                     "bad format",
+			expectedStatus:             404,
+			enableTransactionSimulator: false,
+		},
+		{
+			txnIndex:                   0,
+			format:                     "json",
+			expectedStatus:             200,
+			enableTransactionSimulator: true,
+		},
+		{
+			txnIndex:                   0,
+			format:                     "msgpack",
+			expectedStatus:             200,
+			enableTransactionSimulator: true,
+		},
+		{
+			txnIndex:                   0,
+			format:                     "bad format",
+			expectedStatus:             400,
+			enableTransactionSimulator: true,
+		},
+	}
+
+	for i, testCase := range testCases {
+		testCase := testCase
+		t.Run(fmt.Sprintf("i=%d", i), func(t *testing.T) {
+			t.Parallel()
+			simulateTransactionTest(t, testCase.txnIndex, testCase.format, testCase.expectedStatus, testCase.enableTransactionSimulator)
+		})
+	}
+}
+
+func copyInnerTxnGroupIDs(t *testing.T, dst, src *model.PendingTransactionResponse) {
+	t.Helper()
+
+	// msgpack decodes to map[interface{}]interface{} while JSON decodes to map[string]interface{}
+	txn := dst.Txn["txn"]
+	switch dstTxnMap := txn.(type) {
+	case map[string]interface{}:
+		srcTxnMap := src.Txn["txn"].(map[string]interface{})
+		groupID, hasGroupID := srcTxnMap["grp"]
+		if hasGroupID {
+			dstTxnMap["grp"] = groupID
+		}
+	case map[interface{}]interface{}:
+		srcTxnMap := src.Txn["txn"].(map[interface{}]interface{})
+		groupID, hasGroupID := srcTxnMap["grp"]
+		if hasGroupID {
+			dstTxnMap["grp"] = groupID
+		}
+	}
+
+	if dst.InnerTxns == nil || src.InnerTxns == nil {
+		return
+	}
+
+	assert.Equal(t, len(*dst.InnerTxns), len(*src.InnerTxns))
+
+	for innerIndex := range *dst.InnerTxns {
+		if innerIndex == len(*src.InnerTxns) {
+			break
+		}
+		dstInner := &(*dst.InnerTxns)[innerIndex]
+		srcInner := &(*src.InnerTxns)[innerIndex]
+		copyInnerTxnGroupIDs(t, dstInner, srcInner)
+	}
+}
+
+func assertSimulationResultsEqual(t *testing.T, expectedError string, expected, actual model.SimulationResponse) {
+	t.Helper()
+
+	if len(expectedError) != 0 {
+		require.NotNil(t, actual.TxnGroups[0].FailureMessage)
+		require.Contains(t, *actual.TxnGroups[0].FailureMessage, expectedError)
+		require.False(t, expected.WouldSucceed, "Test case WouldSucceed value is not consistent with expected failure")
+		// if it matched the expected error, copy the actual one so it will pass the equality check below
+		expected.TxnGroups[0].FailureMessage = actual.TxnGroups[0].FailureMessage
+	}
+
+	// Copy inner txn groups IDs, since the mocktracer scenarios don't populate them
+	assert.Equal(t, len(expected.TxnGroups), len(actual.TxnGroups))
+	for groupIndex := range expected.TxnGroups {
+		if groupIndex == len(actual.TxnGroups) {
```

Comment:
Good question. These aren't explicitly necessary, but I think they make error messages clearer.

`assert.Equal`, unlike `require.Equal`, won't terminate the test if it fails. I want this behavior so that the test always gets to line 847, which contains a call to `require.Equal` that will helpfully print out the two objects if they're unequal.

The problem is that even with `assert.Equal`, the test will still fail in the loop if `actual.TxnGroups` is shorter than `expected.TxnGroups`, so I introduced the break statement.

Reply:
That makes sense - thanks for the explanation 👍